### DISCUSSION TITLE
Recursively set Config setting "extra" to "forbid" or "ignore"

### DISCUSSION THREAD

**Discussion by daviskirk** [pydantic/pydantic#2652] [Feature Request]
This might be a feature request but I thought I'd ask if anyone had a similar problem/use case and perhaps workarounds or solutions before I file an issue:

Is there a way of recursively setting the "extra" argument of a model config?
My use case is a data source that I read from that has additional fields I want to ignore, but at the same time for input data I want to forbid these "extra" data.
Example would be something like this:

```python
from pydantic import BaseModel

class A(BaseModel):
    foo: str = "foo"

class B(BaseModel):
    a: A = A()

# when reading from data source:

B.parse_obj({"a": {"foo": "bar"}, "ignored": "ignored value"})
# B(a=A(foo='bar'))

# Now when getting input we want to forbid additional fields:

class C(B):
    class Config:
        extra = "forbid"

C.parse_obj({"a": {"foo": "bar"}, "bad": "input"})
# This works:
# ValidationError: 1 validation error for C
# bad
#   extra fields not permitted (type=value_error.extra)

# but this does not:
C.parse_obj({"a": {"foo": "bar", "bad": "input"}})
# C(a=A(foo='bar'))
```

Anyone have any way of doing this without going through massive gymnastics of parsing all fields and recursively subclassing all field classes with the "forbid" extra setting?

**Comment by falknerdominik** [comment]
Hi,
not a dev but you could preconfigure a base class.
```
class ConfiguredBaseModel(BaseModel):
    class Config:
        extra = 'forbid'


class B(ConfiguredBaseModel):
    ...

# your code
```

**Reply by adriangb** [reply]
This is the way to do it if you don't mind having it set to one thing for all models.
If you sometimes want to use models with `extra='forbid'` and other times with `extra='ignore'` or something like that it's a bit trickier and we don't currently support it. What we've done so far is make these things runtime configurable, e.g. the `strict` parameter. @samuelcolvin what do you think of making `extra` runtime configurable as well?

**Reply by zmievsa** [reply]
@adriangb According to Samuel in his post below, he did make extra runtime configurable. Or was that a misunderstanding?

**Reply by samuelcolvin** [reply]
No, I said we made config recursive, but we had to remove that as it caused errors.

@adriangb I'm not opposed to making `extra` optionally recursive.

I'm a bit less sure about making it runtime configurable as I imagine knowing it at build time might allow optimisations in the future.

**Comment by evstratbg** [comment]
I have the same need
@daviskirk did you resolve it?

**Comment by zmievsa** [comment]
@daviskirk @evstratbg I made https://github.com/Ovsyanka83/pydantic-duality specifically for this use case. It solves the problem in its entirety for me.

**Comment by samuelcolvin** [comment]
Pydantic v2 allows config to be set recursively, it's already implemented in pydantic-core, we just need to document it.

**Reply by zmievsa** [reply]
@samuelcolvin would you please describe how it works or point me to the place where this functionality is defined? Seems like you split Config and Schema, thus allowing us to apply a config from the top schema to all nested schemas during validation. But I can't find **how** to do that or how it actually works.

I hope to make pydantic-duality somewhat compatible with pydanticV2's interface to make sure the transition for my users is going to be easy.

**Reply by mfisher87** [reply]
>But I can't find how to do that

I'm having the same problem. I've been looking for a long time and have found second-hand evidence that this is possible, like this thread, but feel really dumb because I haven't been able to achieve it.

**Reply by mfisher87** [reply]
I eventually found the documentation page I'm looking for: https://docs.pydantic.dev/latest/concepts/config/#change-behaviour-globally

It moved since <=2.3 (https://docs.pydantic.dev/2.3/usage/model_config/#change-behaviour-globally), which was a large part of my problem.

**Reply by Bernardoow** [reply]
@mfisher87  heey !!!

Is This doc the undocument thing that @samuelcolvin told us?

**Reply by mfisher87** [reply]
Yes, I believe so!

**Comment by felixfontein** [comment]
I've also stumbled about this problem. My use-case is to do parsing and linting with the same structure: for parsing it is OK if there are extra fields, but the linter should flag them. So when the linting step is run, I need to be able to switch `extra` from `ignore` to `forbid`.

In pydantic v1 it was easy to do this with `cls.__config__.extra = Extra.forbid`, that was applied recursively (assuming you never set it explicitly anywhere in the sub-models). But with v2 you have to do `cls.model_config["extra"] = "forbid"`, call `cls.model_rebuild(force=True)`, and make sure that already did the same for all models used in attributes before calling `model_rebuild`.

I wrote some code to automate that: https://github.com/ansible-community/antsibull-core/pull/169/files#diff-76b12ca8d8b3847c4b1c5c5bac41685fb6131a0e52b3620e9a8836e5372f1003 This seems to work fine for my classes, but obviously 

It would be a lot easier if pydantic would make this easier though, for example by allowing to pass `extra=Extra.forbid` to `cls.model_validate()`, or if `extra` would be optionally recursive so I only have to modify the base class' config and call its `model_rebuild()`, instead of having to go through all models it directly or indirectly use and update them in the right order.

**Comment by r4victor** [comment]
This functionality is super important for https://github.com/dstackai/dstack. We have lots of pydantic models defined, and we need to parse with extra="forbid" when parsing user input (e.g. configuration files or API requests). On the other hand, we need to parse with extra="ignore" when parsing server responses on the client (for backward compatibility) or parsing records from the DB.

So far we've been using https://github.com/zmievsa/pydantic-duality which is great but no longer maintained. We'd like to have this functionality in pydantic, i.e. to be able to parse models with extra="forbid" or extra="ignore" without defining the entire hierarchy of models twice.

IMHO, having this in pydantic core makes a lot of sense since it's basically a must-have for projects that include both client library and server side and have production requirements on error reporting, backward compatibility, etc.

**Comment by lowdewijk** [comment]
Here is an automatic "massive gymnastics" workaround

```python
from __future__ import annotations
from typing import (
    Annotated,
    Any,
    TypeVar,
    get_args,
    get_origin,
    Union,
    cast,
    Mapping,
    MutableMapping,
)
from pydantic import BaseModel, ConfigDict

T = TypeVar("T", bound=BaseModel)


def set_model_config_recursively(cls: type[T], config: ConfigDict) -> type[T]:
    """
    This function creates a subclass of the given Pydantic BaseModel class `cls`,
    and sets the `model_config` attribute to the given `config`. If any fields of
    `cls` are themselves BaseModel subclasses, those are also replaced with
    subclasses with the given `config`, recursively.

    Use-case: The only way to let Pydantic know that we want to have errors when
    extra fields are present is by setting a static field on each model class:

    ```
    model_config = ConfigDict(extra="ignore")
    ```

    If however, you want to do this dynamically, e.g. you want to be strict some of
    the time and lenient at other times, you essentially need another copy of the
    model class with a different config. If the model contains other models as fields,
    those also need to be replaced with copies with the new config, and so on.
    This function does exactly that, but at runtime, not coding time.

    This is a workaround for a missing feature in Pydantic:
    https://github.com/pydantic/pydantic/discussions/2653

    Args:
        cls: The BaseModel subclass to set model_config on.
    Returns:
        A subclass of `cls` with the given config, and any BaseModel
        fields replaced with subclasses with the given config.
    """

    def _recurse_annotation(ann: Any) -> Any:
        # Swap only BaseModel types; leave everything else untouched
        try:
            if issubclass(ann, BaseModel):
                return set_model_config_recursively(cast(type[BaseModel], ann), config)
        except TypeError:
            pass

        origin = get_origin(ann)
        if origin is None:
            return ann

        args = get_args(ann)

        # Optional/Union
        if origin is Union and args:
            new_args = tuple(_recurse_annotation(a) for a in args)
            return Union[new_args]  # type: ignore[index]

        # Annotated[T, ...] — preserve metadata
        if origin is Annotated and args:
            base = _recurse_annotation(args[0])
            meta = args[1:]
            return Annotated[base, *meta]  # type: ignore[misc]

        # list[T], set[T], frozenset[T]
        if origin in (list, set, frozenset) and args:
            inner = _recurse_annotation(args[0])
            return origin[inner]  # type: ignore[index]

        # tuple[T, ...] or tuple[T1, T2, ...]
        if origin is tuple and args:
            if len(args) == 2 and args[1] is Ellipsis:
                inner = _recurse_annotation(args[0])
                return tuple[inner, ...]  # type: ignore[index]
            else:
                return tuple[tuple(_recurse_annotation(a) for a in args)]  # type: ignore[index]

        # dict[K, V] or Mapping[K, V]
        if origin in (dict, Mapping, MutableMapping) and len(args) == 2:
            k, v = args
            return dict[_recurse_annotation(k), _recurse_annotation(v)]  # type: ignore[index]

        return ann

    # build a subclass that only overrides config
    name = f"Lenient{cls.__name__}"
    attrs: dict[str, Any] = {
        "model_config": config,
    }

    # only rewrite annotations for fields whose type is (or contains) BaseModel
    anns: dict[str, Any] = {}
    for field_name, f in cls.model_fields.items():
        ann = f.annotation
        new_ann = _recurse_annotation(ann)
        if new_ann is not ann:
            anns[field_name] = new_ann  # shadow only when changed

    if anns:
        attrs["__annotations__"] = anns

    UpdatedClass = cast(type[T], type(name, (cls,), attrs))
    # rebuild schema so overridden annotations (if any) take effect
    UpdatedClass.model_rebuild()
    return UpdatedClass
```


I use it to set `extra="ignore"`, since I have already `extra="forbid"` as the default of  my Pydantic base class

```python
config = set_model_config_recursively(MyConfig, ConfigDict(extra="ignore")).model_validate(my_config_dict)
```
Note: this won't handle self-recursion in model classes. For that you need to build a cache to break the self-recursive cycle.

**Comment by r4victor** [comment]
**This is supported since Pydantic 2.12!**

> ## extra configuration per validation
> It is now possible to control the [extra](https://pydantic.dev/docs/validation/2.12/concepts/models/#extra-data) behavior per validation call:
>
> 
> ```python
> from pydantic import BaseModel
> 
> 
> class Model(BaseModel):
>     x: int
> 
>     model_config = ConfigDict(extra='allow')
> 
> 
> # Validates fine:
> m = Model(x=1, y='a')
> # Raises a validation error:
> m = Model.model_validate({'x': 1, 'y': 'a'}, extra='forbid')
> ```

https://pydantic.dev/articles/pydantic-v2-12-release

**Reply by felixfontein** [reply]
Thanks for pointing that out! Link to PRs: https://github.com/pydantic/pydantic/pull/12233 / https://github.com/pydantic/pydantic-core/pull/1722
